src/object/cli_obj.c

```diff
@@ -3312,13 +3326,14 @@ obj_punch_common(tse_task_t *task, enum obj_rpc_opc opc, daos_obj_punch_t *args)
 
 	if (daos_handle_is_valid(args->th)) {
 		/* add the operation to DTX and complete immediately */
-		rc = dc_tx_attach(args->th, args, opc);
+		rc = dc_tx_attach(args->th, args, opc, task);
```

Comment:
We could also complete task inside dc_tx_attach and just do the follow here:

    return dc_tx_attach(..., task);

Reply:
> We could also complete task inside dc_tx_attach and just do the follow here:
> 
> ```
> return dc_tx_attach(..., task);
> ```

It is better not complete the task inside dc_tx_attach(), because we will have another patch to cache read operation that is called via fetch complete callback.